sky/engine/bindings/jni/dart_jni.cc

```diff
@@ -112,40 +188,133 @@ JniClass::JniClass(JNIEnv* env, jclass clazz)
 JniClass::~JniClass() {
 }
 
-PassRefPtr<JniClass> JniClass::fromName(const char* name) {
-  JNIEnv* env = base::android::AttachCurrentThread();
+PassRefPtr<JniClass> JniClass::FromName(const char* name) {
+  Dart_Handle exception = NULL;
+  {
+    ENTER_JNI();
 
-  ScopedJavaLocalRef<jclass> clazz = DartJni::GetClass(env, name);
-  CheckJniException(env);
+    ScopedJavaLocalRef<jclass> clazz = DartJni::GetClass(env, name);
+    if (CheckJniException(env, &exception)) goto fail;
 
-  return adoptRef(new JniClass(env, clazz.obj()));
+    return adoptRef(new JniClass(env, clazz.obj()));
+  }
+fail:
+  Dart_ThrowException(exception);
+  ASSERT_NOT_REACHED();
 }
 
-intptr_t JniClass::getFieldId(const char* name, const char* sig) {
-  JNIEnv* env = base::android::AttachCurrentThread();
-  jfieldID id = env->GetFieldID(clazz_.obj(), name, sig);
-  CheckJniException(env);
-  return reinterpret_cast<intptr_t>(id);
+intptr_t JniClass::GetFieldId(const char* name, const char* sig) {
+  Dart_Handle exception = NULL;
+  {
+    ENTER_JNI();
+
+    jfieldID id = env->GetFieldID(clazz_.obj(), name, sig);
+    if (CheckJniException(env, &exception)) goto fail;
+
+    return reinterpret_cast<intptr_t>(id);
+  }
+fail:
+  Dart_ThrowException(exception);
+  ASSERT_NOT_REACHED();
```

Comment:
Woah, you're right.  Ok, we need to look at all our uses of Dart_ThrowException because that's not what I thought it did and I'm sure we have lots of bugs because of that.  :(